### PR TITLE
check-esxi-hardware: refactor from python2 to python3

### PR DESCRIPTION
**Pull request by freezeboy**
###### Motivation for this change

Working on issue: #101964 

I took the opportunity to update the dependencies to support python3

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 102282` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>check-esxi-hardware</li>
    <li>python27Packages.nocasedict</li>
    <li>python27Packages.nocaselist</li>
    <li>python27Packages.pywbem</li>
    <li>python27Packages.yamlloader</li>
    <li>python37Packages.nocasedict</li>
    <li>python37Packages.nocaselist</li>
    <li>python37Packages.yamlloader</li>
    <li>python38Packages.nocasedict</li>
    <li>python38Packages.nocaselist</li>
    <li>python38Packages.pywbem</li>
    <li>python38Packages.yamlloader</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 102282` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>check-esxi-hardware</li>
    <li>python27Packages.nocasedict</li>
    <li>python27Packages.nocaselist</li>
    <li>python27Packages.pywbem</li>
    <li>python27Packages.yamlloader</li>
    <li>python37Packages.nocasedict</li>
    <li>python37Packages.nocaselist</li>
    <li>python37Packages.yamlloader</li>
    <li>python38Packages.nocasedict</li>
    <li>python38Packages.nocaselist</li>
    <li>python38Packages.pywbem</li>
    <li>python38Packages.yamlloader</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 102282` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>check-esxi-hardware</li>
    <li>python27Packages.nocasedict</li>
    <li>python27Packages.nocaselist</li>
    <li>python27Packages.pywbem</li>
    <li>python27Packages.yamlloader</li>
    <li>python37Packages.nocasedict</li>
    <li>python37Packages.nocaselist</li>
    <li>python37Packages.yamlloader</li>
    <li>python38Packages.nocasedict</li>
    <li>python38Packages.nocaselist</li>
    <li>python38Packages.pywbem</li>
    <li>python38Packages.yamlloader</li>
  </ul>
</details>

File: pkgs/servers/monitoring/plugins/esxi.nix [hunk 1]
```diff
@@ -27,7 +27,7 @@ in python2Packages.buildPythonApplication rec {
     runHook postInstall
   '';
 
-  propagatedBuildInputs = with python2Packages; [ pywbem ];
+  propagatedBuildInputs = with python3Packages; [ pywbem requests ];
```
Comment (jonringer) [inline pkgs/servers/monitoring/plugins/esxi.nix:30]:
```suggestion
  propagatedBuildInputs = with python3Packages; [ pywbem requests setuptools ];
```

```
[nix-shell:~/.cache/nixpkgs-review/pr-102282]$ ./results/check-esxi-hardware/bin/check_esxi_hardware.py --help
Traceback (most recent call last):
  File "/nix/store/fx2jz12qvfvv3q4ckwqbzyrfgis67sdz-check-esxi-hardware-20200710/bin/.check_esxi_hardware.py-wrapped", line 290, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'
```

Comment (SuperSandro2000) [reply pkgs/servers/monitoring/plugins/esxi.nix:30]: Lets take this opportunity to add a simple test for that:

```
  checkPhase = ''
    $out/bin/check_esxi_hardware.py --help
  '';
```

Comment (FRidh) [reply pkgs/servers/monitoring/plugins/esxi.nix:30]: `setuptools` is a default build dependency, so this test will always succeed during a build, regardless of whether `setuptools` is available during runtime outside a build. Note this is also a bit of an issue with `pythonImportsCheck` and all tests.

File: pkgs/servers/monitoring/plugins/esxi.nix [hunk 2]
```diff
@@ -1,18 +1,18 @@
-{ stdenv, fetchFromGitHub, python2Packages }:
+{ stdenv, fetchFromGitHub, python3Packages }:
 
 let
   bName = "check_esxi_hardware";
   pName = stdenv.lib.replaceStrings [ "_" ] [ "-" ] bName;
 
-in python2Packages.buildPythonApplication rec {
+in python3Packages.buildPythonApplication rec {
```
Comment (FRidh) [inline pkgs/servers/monitoring/plugins/esxi.nix:7]: There is an executable that is installed that has an extension, please remove the extension. Executables in `/bin` should not have any.

File: pkgs/servers/monitoring/plugins/esxi.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/servers/monitoring/plugins/esxi.nix]: you might be able to do something like:
```
  checkPhase = ''
    unset PYTHONPATH
    $out/bin/check_esxi_hardware.py --help
  '';
```

**Comment by peterhoeg** [comment]
I really want to try this out *but* we have this check deployed on *many* locations so I need to be sure it doesn't break anything. Has anyone tried this against actual ESXi instances?

**Comment by jonringer** [comment]
@peterhoeg 
```
NIX_PATH=nixpkgs=https://github.com/freezeboy/nixpkgs/archive/update-check-esxi-hardware.tar.gz nix-shell -p check-esxi-hardware
```

**Comment by peterhoeg** [comment]
Thanks @jonringer  - it wasn't so much about "how" to try it out but more if anyone had actually given this a go.

**Comment by SuperSandro2000** [comment]
@freezeboy Can you resolve the merge conflict?

**Review by jonringer — APPROVED** [review]
LGTM
shows usage

```
https://github.com/NixOS/nixpkgs/pull/102282
13 packages built:
check-esxi-hardware python27Packages.nocasedict python27Packages.nocaselist python27Packages.pywbem python27Packages.yamlloader python37Packages.nocasedict python37Packages.nocaselist python37Packages.pywbem python37Packages.yamlloader python38Packages.nocasedict python38Packages.nocaselist python38Packages.pywbem python38Packages.yamlloader
```

**Comment by jonringer** [comment]
> Thanks @jonringer - it wasn't so much about "how" to try it out but more if anyone had actually given this a go.

if there was an issue with the interpreter, it's usually very apparent

**Pull request merged**